src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7363,9 +7363,19 @@ private BoundExpression ConvertToArrayIndex(BoundExpression index, SyntaxNode no
 
             var result =
                 TryImplicitConversionToArrayIndex(index, SpecialType.System_Int32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_Int64, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt64, node, diagnostics);
+                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics);
+
+            if (result is null && allowNIntAndNUInt)
+            {
+                result = TryImplicitConversionToArrayIndex(index, SpecialType.System_IntPtr, node, diagnostics) ??
+                         TryImplicitConversionToArrayIndex(index, SpecialType.System_UIntPtr, node, diagnostics);
```

Comment:
It looks like this violates C# language specification. Perhaps this should be implemented as an IL gen optimization instead?

Reply:
Is it worth updating the language spec to allow this and if not, where is the right place to put in IL gen optimizations/is there any concern with folding away the `.ovf` instruction currently inserted?

Reply:
> Is it worth updating the language spec to allow this 

It looks to me that the C# code is allowed today, i.e. there is no real need to allow anything new from the language point of view. Actual IL generated is outside of the language specification.

---
In reply to: [567999492](https://github.com/dotnet/roslyn/pull/50910#discussion_r567999492) [](ancestors = 567999492)

Reply:
> It looks to me that the C# code is allowed today, i.e. there is no real need to allow anything new from the language point of view. Actual IL generated is outside of the language specification.

To clarify, I do not think there is a need to make any changes to the binding.

---
In reply to: [568003318](https://github.com/dotnet/roslyn/pull/50910#discussion_r568003318) [](ancestors = 568003318,567999492)